libfreerdp/codec/h264_ffmpeg.c

```diff
@@ -141,7 +178,11 @@ static BOOL libavcodec_create_encoder(H264_CONTEXT* WINPR_RESTRICT h264)
 		return TRUE;
 
 	libavcodec_destroy_encoder(h264);
+#ifdef WITH_VAAPI_H264_ENCODING
```

Comment:
Here perhaps you should instead have a fallback mechanism, like we look for h264_vaapi if the option is set and if not found we go back the default implementation...

Reply:
For encoding, providing run time fallback to software encoder (if hardware initialization fails) will require complete new API, including entry points to `h264_ffmpeg` and `h264_openh264` subsystems. Main reason for this is that API user needs to be notified back that fallback happened during initialization in order to feed YUV420P frames to the encoder (instead of NV12).

With currently proposed change, other h264 encoder API users (if any) will fail if `WITH_VAAPI_H264_ENCODING` is defined, because they are unaware they need to feed NV12 frames. My guess is that xfreerdp client today doesn't have any other users for encoder APIs. The server might be the only other user and needs to be compiled without `WITH_VAAPI_H264_ENCODING`.

Another needed API change will be to provide frame size (width, height) to `h264_context_new()`, because hardware encoder context initialization requires width and height. Today context initialization only happens when `h264->subsystem->Compress()` is called for the 1st time, but it's too late to notify the user which frames (YUV420P or NV12) to feed at this point.

Hope this adds some useful info and not confuses you further :)

Reply:
@oleg0421 see my comments above.
you are right that you need to change something, but it is sufficient to fix the new API here by just providing a source buffer and which format it used.

the extension for the `H264_CONTEXT_SUBSYSTEM` is required, yes, but that is backward compatible (just return an error if the function pointer is not initialized)